apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/MSALoginActivity.java

```diff
@@ -0,0 +1,396 @@
+        }
+    }
+
+    private void registerAppCenterAuthentication(String userId) {
+        AuthenticationProvider.TokenProvider tokenProvider = new AuthenticationProvider.TokenProvider() {
+
+            @Override
+            public void acquireToken(String ticketKey, AuthenticationProvider.AuthenticationCallback callback) {
+
+                /* Refresh token, doing that even on first time to test the refresh code without having to wait 1 hour. */
+                refreshToken(callback);
+            }
+        };
+        AuthenticationProvider provider = new AuthenticationProvider(mAuthType, userId, tokenProvider);
+        AnalyticsTransmissionTarget.addAuthenticationProvider(provider);
+        finish();
+        Toast.makeText(this, R.string.signed_in, Toast.LENGTH_SHORT).show();
+    }
+
+    /**
+     * Get initial access token.
+     */
+    private void getToken(final String code) {
+        Map<String, String> headers = new HashMap<>();
+        headers.put(DefaultHttpClient.CONTENT_TYPE_KEY, "application/x-www-form-urlencoded");
+        sHttpClient.callAsync(TOKEN_URL,
+                DefaultHttpClient.METHOD_POST,
+                headers,
+                new HttpClient.CallTemplate() {
+
+                    @Override
+                    public String buildRequestBody() {
+                        return REDIRECT_URI_PARAM +
+                                CLIENT_ID_PARAM +
+                                "&grant_type=authorization_code" +
+                                "&code=" + code;
+                    }
+
+                    @Override
+                    public void onBeforeCalling(URL url, Map<String, String> headers) {
+                        AppCenterLog.verbose(AppCenter.LOG_TAG, "Calling " + url + "...");
+                    }
+                },
+                new ServiceCallback() {
+
+                    @Override
+                    public void onCallSucceeded(HttpResponse httpResponse) {
+                        try {
+                            JSONObject response = new JSONObject(httpResponse.getPayload());
+                            String userId = response.getString(USER_ID);
+                            mRefreshToken = response.getString(REFRESH_TOKEN);
+                            mRefreshTokenScope = response.getString(SCOPE);
+                            registerAppCenterAuthentication(userId);
+                        } catch (JSONException e) {
+                            onCallFailed(e);
+                        }
+                    }
+
+                    @Override
+                    public void onCallFailed(Exception e) {
+                        handleCallFailure(e);
+                    }
+                });
+    }
+
+    private void refreshToken(final AuthenticationProvider.AuthenticationCallback callback) {
+        Map<String, String> headers = new HashMap<>();
+        headers.put(DefaultHttpClient.CONTENT_TYPE_KEY, "application/x-www-form-urlencoded");
```

Comment:
const?

Reply:
Thanks for the feedback but we can clean it up in another task. This is just copied from another place and we don't change any lines of code here.